explainer-beforematch.md

```diff
@@ -191,6 +191,86 @@ In general, we believe that the `beforematch` event does not pose a privacy prob
+location to scroll to is different.
+3. **display: none**: The `beforematch` event handler adds the style
+`display: none` to the target element.
+4. **visibility: hidden**: The `beforematch` event handler adds the style
+`visibility: hidden` to the target element.
+5. **subtree-visibility: hidden-matchable -> visible**: The target element has
+the style `subtree-visibility: hidden-matchable` before the `beforematch` event
+is fired on the target element, and the `beforematch` event handler changes the
+style value from `hidden-matchable` to `visible`.
+
+
+Here is the current scrolling behavior in chromium for each of those cases in
+ScrollToTextFragment, and ElementFragment, and find-in-page:
+
+#### ScrollToTextFragment
+1. **Removed from dom**: ScrollToTextFragment will not scroll to the removed
+element. If there is a second match which is still in the dom after
+beforematch, it will be scrolled to.
+2. **Reparented in dom**: ScrollToTextFragment will scroll to the new
+location of the target. If there is a second match which was not reparented,
+ScrollToTextFragment will scroll to whichever comes first after the beforematch
+event is fired.
+3. **display: none**: ScrollToTextFragment will not scroll to the target
+element. If there is a second match which was not modified, ScrollToTextFragment
+will scroll to it.
+4. **visibility: hidden**: ScrollToTextFragment will not scroll to the target
+element. If there is a second match which was not modified, ScrollToTextFragment
+will scroll to it.
+5. **subtree-visibility: hidden-matchable -> visible**: ScrollToTextFragment
+will scroll to the unlocked text.
+
+#### ElementFragment
+This is the behavior when navigating to a page with an element fragment, rather
+than the fragment being set by script.
+1. **Removed from dom**: The page will not scroll.
+2. **Reparented in dom**: The page will scroll to the target element's new
+location.
+3. **display: none**: The page will not scroll.
+4. **visibility: hidden**: The page will scroll to the target element. This
+behavior makes sense because it matches the behavior when targeting a
+`visibility: hidden` element with a fragment regardless of beforematch.
+5. **subtree-visibility: hidden-matchable -> visible**: The page will scroll
+to the activated element.
+
+#### find-in-page
+The find-in-page behavior has some issues in these cases. These will likely
+be fixed after adding an async step to pick up layout changes made by the
+beforematch event listener.
+1. **Removed from dom**: find-in-page does not scroll to the target element. If
+there is a second match in the DOM, find-in-page will immediately find it and
+scroll to it.
+2. **Reparented in dom**: If the only match in the document is reparented to the
+end of the document, find-in-page becomes stuck at "0/x" matches and can't
+scroll to the new location, even if you keep typing more of the match out or
+close and reopen find-in-page. This behavior is obviously not good, and
```

Comment:
> This behavior is obviously not good

Yep, we should probably fix it. 

There's references to an 'async step' here. Could you file a crbug and reference it here, just so that we can discuss and track progress?

Reply:
I made a bug: https://bugs.chromium.org/p/chromium/issues/detail?id=1074121
and linked it at the text under ####find-in-page